jdaviz/configs/cubeviz/plugins/parsers.py

```diff
@@ -88,36 +88,84 @@ def parse_data(app, file_obj, data_type=None, data_label=None):
         raise NotImplementedError(f'Unsupported data format: {file_obj}')
 
 
+def _return_spectrum_with_correct_units(flux, wcs, metadata, data_type, target_wave_unit=None,
+                                        hdulist=None):
+    """Upstream issue of WCS not using the correct units for data must be fixed here.
+    Issue: https://github.com/astropy/astropy/issues/3658
+    """
+    with warnings.catch_warnings():
+        warnings.filterwarnings(
+            'ignore', message='Input WCS indicates that the spectral axis is not last',
+            category=UserWarning)
+        sc = Spectrum1D(flux=flux, wcs=wcs)
+
+    # TODO: Can the unit be defined in a different keyword, e.g., CUNIT1?
+    if target_wave_unit is None and hdulist is not None:
+        cunit_key = 'CUNIT3'
+        for ext in ('SCI', 'FLUX', 'PRIMARY'):  # In priority order
+            if ext in hdulist and cunit_key in hdulist[ext].header:
+                target_wave_unit = u.Unit(hdulist[ext].header[cunit_key])
+                break
+
+    if (data_type == 'flux' and target_wave_unit is not None
+            and target_wave_unit != sc.spectral_axis.unit):
+        metadata['_orig_wcs'] = wcs
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                'ignore', message='Input WCS indicates that the spectral axis is not last',
+                category=UserWarning)
+            new_sc = Spectrum1D(
+                flux=sc.flux,
+                spectral_axis=sc.spectral_axis.to(target_wave_unit, u.spectral()),
+                meta=metadata)
+    else:
+        sc.meta = metadata
+        new_sc = sc
+    return new_sc
+
+
 def _parse_hdulist(app, hdulist, file_name=None):
-    if file_name is None:
-        if hasattr(hdulist, 'file_name'):
-            file_name = hdulist.file_name
+    if file_name is None and hasattr(hdulist, 'file_name'):
+        file_name = hdulist.file_name
+    else:
+        file_name = file_name or "Unknown HDU object"
 
-    file_name = file_name or "Unknown HDU object"
+    is_loaded = []
+    wcs_sci = None
 
-    # Now loop through and attempt to parse the fits extensions as spectral
-    #  cube object. If the wcs fails to parse in any case, use the wcs
-    #  information we scraped above.
+    # TODO: This needs refactoring to be more robust.
+    # Current logic fails if there are multiple EXTVER.
     for hdu in hdulist:
-        data_label = f"{file_name}[{hdu.name}]"
-
         if hdu.data is None or not hdu.is_image or hdu.data.ndim != 3:
             continue
 
-        try:
+        data_type = _get_data_type_by_hdu(hdu)
+        if not data_type:
+            continue
+
+        # Only load each type once.
+        if data_type in is_loaded:
+            continue
```

Comment:
so this loads the _first_ instance (in `hdulist`) of each of the data types?  How common are additional instances?  Should we raise snackbar warnings or just silently ignore those?

Reply:
this is probably fine for now, and if anyone complains, we can add a snackbar warning.  I hate to add extra code if its a use-case that won't ever actually happen.